realsense_camera/src/base_nodelet.cpp

```diff
@@ -275,34 +276,67 @@ namespace realsense_camera
   {
     // print list of detected cameras
     std::vector<int> camera_type_index;
-    std::string detected_camera_msg = " - Detected the following cameras:";
+
+    // map camera names to validated firmware
+    std::map<std::string, std::string> camera_name_to_validated_firmware =
+    {
+          {"Intel RealSense R200", R200_CAMERA_FW_VERSION},
+          {"Intel RealSense F200", F200_CAMERA_FW_VERSION},
+          {"Intel RealSense SR300", SR300_CAMERA_FW_VERSION},
+          {"Intel RealSense ZR300", ZR300_CAMERA_FW_VERSION},
+          {"Intel RealSense ZR300_ADAPTER", ZR300_ADAPTER_FW_VERSION},
+          {"Intel RealSense ZR300_MOTION_MODULE", ZR300_MOTION_MODULE_FW_VERSION}
+    };
+
     for (int i = 0; i < num_of_cameras; i++)
     {
+      std::string detected_camera_msg = " - Detected the following camera:";
+      std::string warning_msg = " - Detected unvalidated firmware:";
       // get device
       rs_device* rs_detected_device = rs_get_device(rs_context_, i, &rs_error_);
 
+      // get camera serial number
+      std::string camera_serial_number = rs_get_device_serial(rs_detected_device, &rs_error_);
+      checkError();
+
       // get camera name
       std::string camera_name = rs_get_device_name(rs_detected_device, &rs_error_);
       checkError();
 
+      // get camera firmware
+      std::string camera_fw = rs_get_device_firmware_version(rs_detected_device, &rs_error_);
+      checkError();
+
       if (camera_name.find(camera_type_) != std::string::npos)
       {
         camera_type_index.push_back(i);
       }
       // print camera details
       detected_camera_msg = detected_camera_msg +
-            "\n\t\t\t\t- Serial No: " + rs_get_device_serial(rs_detected_device, &rs_error_) +
-            ", USB Port ID: " + rs_get_device_usb_port_id(rs_detected_device, &rs_error_) +
+            "\n\t\t\t\t - Serial No: " + camera_serial_number + ", USB Port ID: " +
+            rs_get_device_usb_port_id(rs_detected_device, &rs_error_) +
             ", Name: " + camera_name +
-            ", Camera FW: " + rs_get_device_firmware_version(rs_detected_device, &rs_error_);
+            ", Camera FW: " + camera_fw;
       checkError();
 
+      if (camera_fw != camera_name_to_validated_firmware[camera_name])
```

Comment:
There's a lot of code duplication here, can the checks of the fw and printing warnings be refactored into a common method that is called for each fw?